### ISSUE TITLE
TypeGuard on TypedDict Change

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#7161]
**Describe the bug**

```python
from typing import TypeGuard, TypedDict

def is_typeddict(t: object) -> TypeGuard[type[TypedDict]]:
    return isinstance(tp, type) and issubclass(tp, dict) and hasattr(tp, "__total__")

def f(x: object):
    if is_typeddict(x):
        print(x.__required_keys__)
```

Before 1.1.336 this worked with no errors. After 1.1.337 there is now an error message of `TypedDict can not be used in this context`. Except the type guard still works and is useful. If you hover over `__required_keys__` it is correctly narrowed. If I change typeguard to just bool, I can no longer narrow type[TypedDict] at runtime.

**VS Code extension or command-line**
Both. pyright >=1.1.337 is where error started. 

If this is expected how would you do this (runtime inspection of types)? Just use Any/ignore?

**Comment by erictraut** [comment]
The new behavior is correct, and I don't consider this a bug.

`TypedDict` is a special form that is used as a base class to define a structural type. It's similar to `Protocol`. It is not valid to use it in a `type`. No type is a subtype of `TypedDict`.

I'm not sure I understand your use case, but perhaps it would make sense to define a protocol that includes a `__required_keys__` attribute?

Out of curiosity, why do you want to determine at runtime whether a class is a `TypedDict` class? Are you performing runtime type checking?

**Comment by hmc-cs-mdrissi** [comment]
The library is internal but very similar to [cattrs](https://catt.rs/en/stable/). The library allows defining parsers that serialize/deserialize a python object. To know how to serialize/deserialize the type it relies on examining runtime type annotations and api also passes expected type. So library usage looks like,

```python
class Foo(TypedDict):
  x: list[str]
  y: SomeComplexObj

to_config(Foo, {"x": [1, 2], "y": SomeComplexObj(...)}
```

and then because Foo is a TypedDict it then inspects annotations and finds x as list[str] for json stays same, for y it will lookup is there a parser defined for `SomeComplexObj`. The required_keys/optional_keys are used at parsing time mainly as validation check that data has expected fields. I also have similar code for a few other special `types` like dataclasses and understanding how field behaves for knowing defaults/how to build the object.

A protocol would probably work although intent of logic is very much based on typed dict definition and maybe newer features (proposed __extras__) I'll add corresponding logic for.

**Comment by Pwuts** [comment]
I'm having a similar issue where `typing.is_typeddict(obj)` doesn't narrow the type to `type[TypedDict]`.

This is a problem because e.g. Pydantic has this method:
```py
def create_model_from_typeddict(
    typeddict_cls: Type['TypedDict'],  # type: ignore[valid-type]
    **kwargs: Any,
) -> Type['BaseModel']:
```
so the first argument MUST be a `TypedDict`. Using `is_typeddict` gives runtime certainty and yet does not narrow the type, so I still have to add a `type: ignore[reportArgumentType]` to the invocation.

My use case:
```py
TD = TypeVar("TD", bound=TypedDict)

def validate(type: Type[TD], data: Any) -> TD:
    """Validate untrusted data matches a given TypedDict"""

    if not is_typeddict(type):
        raise TypeError(f'Only TypedDict types are supported, got: {type} instead.')

    # ... object construction & validation ...

    return cast(TD, obj)
```
